--- lua/ge/extensions/core/input/virtualInput.lua
+++ lua/ge/extensions/core/input/virtualInput.lua
@@ -48,7 +48,7 @@
 local function emit(deviceInstance, objType, objectInstance, action, val)
   local mgr = getVirtualInputManager()
   if not mgr then return end
-  mgr:emitEvent('vinput', deviceInstance, objType, objectInstance, action, val)
+  mgr:emitEvent('vinput', deviceInstance, objType, objectInstance, action, val, 0)
 end
 
 M.createDevice = createDevice
